test: migrate tests/02-cnp-specs.sh to Ginkgo

ianvernon opened the pull request:
To expedite the migration of the Kubernetes tests, which are the rate-determining part of the build, migrate over `tests/02-cnp-specs.sh` in a 1:1 manner. This ensures that we support the bookinfo demo policy configuration as part of our CI.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/provision: allow forwarding packets with iptables

Allow forwarding of packets so kube-dns works.

Signed-off by: Ian Vernon <ian@cilium.io>:

Commit:
test/helpers: add new functions

Add the following functions:

* Create: wrapper around kubectl create
* WaitForServiceEndpoints: waits until the specified service contains the specified port
* GetEndpoints: gets specified endpoints from Kubernetes

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: migrate 02-cnp-specs.sh test to Ginkgo

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/k8sT/manifests: add bookinfo demo files

Add policies, K8s pod / service files, etc.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -223,13 +229,55 @@ func (kub *Kubectl) WaitforPods(namespace string, filter string, timeout time.Du
+		data, err := kub.GetEndpoints(namespace, filter).Filter(jsonPath)
+
+		if err != nil {
+			kub.logger.Errorf("could not get service endpoints: %s", err)
```
`kub.logger.WithErr(err).`

ianvernon replied:
Fixed.

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -223,13 +229,55 @@ func (kub *Kubectl) WaitforPods(namespace string, filter string, timeout time.Du
+			kub.logger.Errorf("could not get service endpoints: %s", err)
+			return false
+		}
+		log.Infof("data: %s", data.String())
```
This log message is not clear, and should be `Debugf`

ianvernon replied:
This was leftover from debugging, will remove in next upload.

ianvernon replied:
Fixed.

eloycoto commented on test/k8sT/Services.go:
```diff
@@ -146,4 +147,162 @@ var _ = Describe("K8sServicesTest", func() {
+		shouldConnect := func(srcPod, dst string) {
+			By(fmt.Sprintf("Checking that %s can connect to %s", srcPod, dst))
+			res := kubectl.Exec(fmt.Sprintf(wgetCommand, helpers.KubectlCmd, srcPod, dst))
+			res.ExpectSuccess()
```
Could you add a message in this ExpectSuccess? If fails we need to know from where to where it fails.

ianvernon replied:
Fixed.

eloycoto commented on test/k8sT/Services.go:
```diff
@@ -146,4 +147,162 @@ var _ = Describe("K8sServicesTest", func() {
+		shouldNotConnect := func(srcPod, dst string) {
+			By(fmt.Sprintf("Checking that %s cannot connect to %s", srcPod, dst))
+			res := kubectl.Exec(fmt.Sprintf(wgetCommand, helpers.KubectlCmd, srcPod, dst))
+			res.ExpectFail()
```
The same as should connect a message to say from where to where and why it works when it shouldn't

ianvernon replied:
Fixed.

eloycoto commented on test/k8sT/Services.go:
```diff
@@ -146,4 +147,162 @@ var _ = Describe("K8sServicesTest", func() {
+		var policyPath string
+		var policyCmd string
+
+		if helpers.GetCurrentK8SEnv() != "1.6" {
```
This Policy should have the same name (CNP.yaml) and should be one general, and the version 1.6 into the 1.6/directory. 

kubectl.ManifestGet take care about the version.

I think that one of the goals is to eliminate the If clauses on test ;-)

ianvernon replied:
Fixed.

eloycoto reviewed (changes requested):
Looks good, but I don't liked the `bookinfo` folder in the manifest root folder, I think that we should have all the manifest in one main folder and individual cases for a version in his folder.

aanm commented on test/k8sT/Services.go:
```diff
@@ -146,4 +147,162 @@ var _ = Describe("K8sServicesTest", func() {
+			output, err := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPodK8s1, policyCmd)
+			Expect(err).Should(Not(BeNil()), "policies should be deleted from Cilium: policies found: %s", output)
+		}()
+		defer kubectl.Delete(policyPath)
```
Be aware this is done async, so when you check if all policies were deleted in Cilium we might get test flakes on line 280 `Expect(err).Should(Not(BeNil()), "policies should be deleted from Cilium: policies found: %s", output)`

ianvernon replied:
Good point, I'll add a check for policy wait here in the next upload.

ianvernon replied:
Fixed.

Commit:
helpers: refactor K8s policy import function

* Refactor function which waits for policies to be imported and applied in Cilium
to handle a variety of lifecycle actions in Kubernetes (create, action, delete).
* Also ensure that in new CNP Specs test, that we use this refactored function.

Signed-off by: Ian Vernon <ian@cilium.io>

eloycoto reviewed (approved):
LGTM, thanks
